Item groups for guns and ammo

SunshineDistillery opened the pull request:
This is to move guns and ammo out of item_groups.json into the new item_groups folder. 
It also splits a few big groups, like ammo, into smaller groups such as pistol_ammo, shotgun_ammo, rifle_ammo, etc. And it adds some new groups like archery and archery_ammo. 

In a few places I added ammo types that belong in a category, for completeness, but set them to 0 so they wouldn't spawn in gun stores or civilian areas.  Those special types will get tacked on to military ammo groups or at locations like an lmoe shelter.

chaosvolt commented:
Hmm, curious. o.o

BevapDin commented:
> In a few places I added ammo types that belong in a category, for completeness, but set them to 0 so they wouldn't spawn in gun stores or civilian areas.

It might be good to document this somewhere, maybe in the JSON file itself? People might think the 0 probability is a mistake.

chaosvolt commented:
An alternative would possibly be a comment in the item group itself.

Or doing both might be best, for the sake of insurance against that event. XP

Rivet-the-Zombie commented on data/json/itemgroups/main.json:
```diff
@@ -222,6 +222,629 @@
+  },{
+    "type" : "item_group",
+    "id" : "rifle_ammo",
+    "//" : "some set to 0 so they won't spawn in civilian areas.",
```
>some set to 0 so they won't spawn in civilian areas.

How are the zero weight ones being spawned?

Malkeus commented:
"Those special types will get tacked on to military ammo groups or at locations like an lmoe shelter." Has this been done yet? It would be confusing to lose these spawns while waiting for another PR.

SunshineDistillery commented:
Yeah there's  a comment in the ammo item group . I've tried to put in a lot of comments, maybe more than are really needed. But since these files are used by modders, and maybe first-time modders, I figure it's better to tell too much than too little. 

A lot of those 0 weight items, like .50 cal incendiary rounds, aren't actually in the ammo group right now, so putting them in at weight 0 won't change the current spawns. They are spawned from a bunch of scattered item groups like military, mil_hw, mil_rifles, stash_ammo, swat_gear, or in vehicles. Mostly I just added them at 0 to the ammo lists so that if someone wanted to see all the various types of ammo in the game they could look in one place rather than search all over. 

But I could also make a group called restricted_ammo or something like that. That way if someone wanted to have all ammo types show up in a place, they could easily slap in ammo and restricted_ammo without having to add the missing items in one at a time.

Coolthulhu commented:
Looks like `guns_with_ammo` doesn't work properly with nested item groups yet. This breaks stuff...

Malkeus commented:
Please, please, please, feel free to go comment crazy. I'm over the hump now for json files, but the learning curve is real and it is steeeeeep.

Coolthulhu commented:
From what I can tell, the `guns_with_ammo` flag has to be declared on the lowest level it is used.
This means that the same item group has to either always have all guns with ammo or never have them with ammo.

This behavior probably can't be fixed from the json side.

Look at the LMOE item group:
In old version, it has guns with ammo, meaning that all guns get a single stack of default ammo for them.
In the new version, the guns with ammo flag is declared at "lmoe" group level and then dropped when it recurses into, say, "rifles".

To fix that, you'd either have to duplicate all the groups that are to spawn with ammo (which would be ugly) or wait for someone to add item group flag passing on recursion.

SunshineDistillery commented:
Well, crap. I hadn't noticed that. It looks like there are only a few groups that use guns_with_ammo: lmoe guns, college_camping, home_guns, behind_counter, and some guns in pizzaria trashcans? Out of those, I've only changed lmoe and home guns, but those are pretty important.

I could:  
1> postpone guns til later, and start with other item groups. 
2> make new, temporary item groups for home guns and lmoe guns. 
  2a> or just restore the old item groups for them. 
3> break it and add more ammo to the item groups to make up for it. 
4> wait and hope somebody fixes it.

I'd rather not do 4, because there is so much to get through.  What do you think?

SunshineDistillery commented:
I did a quick test and putting guns_with_ammo at the lowest level worked fine. So I could make duplicate groups rifles_with_ammo, pistols_with_ammo, etc.  as a temporary fix. That would be easy.  Although I understand why "temporary" solutions like that are awful when they might sit around for years.

SunshineDistillery commented:
OK. I've added new item groups that use guns_with_ammo and changed the groups in the locations file, homeguns and lmoe guns, to draw from those. If guns_with_ammo gets updated any time soon, I can come back and remove those redundant categories.
